consensus/blake3pow/consensus.go

```diff
@@ -643,31 +648,282 @@ func (blake3pow *Blake3pow) Finalize(chain consensus.ChainHeaderReader, header *
 		multiSet = rawdb.ReadMultiSet(chain.Database(), header.ParentHash(nodeCtx))
 	}
 	if multiSet == nil {
-		return nil, fmt.Errorf("Multiset is nil for block %s", header.ParentHash(nodeCtx).String())
+		return nil, 0, fmt.Errorf("Multiset is nil for block %s", header.ParentHash(nodeCtx).String())
+	}
+
+	utxoSetSize += uint64(len(utxosCreate))
+	if utxoSetSize < uint64(len(utxosDelete)) {
+		return nil, 0, fmt.Errorf("UTXO set size is less than the number of utxos to delete. This is a bug. UTXO set size: %d, UTXOs to delete: %d", utxoSetSize, len(utxosDelete))
+	}
+	utxoSetSize -= uint64(len(utxosDelete))
+
+	trimDepths := types.TrimDepths
+	if utxoSetSize > params.SoftMaxUTXOSetSize/2 {
+		var err error
+		trimDepths, err = rawdb.ReadTrimDepths(chain.Database(), header.ParentHash(nodeCtx))
+		if err != nil || trimDepths == nil {
+			blake3pow.logger.Errorf("Failed to read trim depths for block %s: %+v", header.ParentHash(nodeCtx).String(), err)
+			trimDepths = make(map[uint8]uint64, len(types.TrimDepths))
+			for denomination, depth := range types.TrimDepths { // copy the default trim depths
+				trimDepths[denomination] = depth
+			}
+		}
+		if UpdateTrimDepths(trimDepths, utxoSetSize) {
+			blake3pow.logger.Infof("Updated trim depths at height %d new depths: %+v", header.NumberU64(nodeCtx), trimDepths)
+		}
+		if !setRoots {
+			rawdb.WriteTrimDepths(batch, header.Hash(), trimDepths)
+		}
+	}
+	start := time.Now()
+	collidingKeys, err := rawdb.ReadCollidingKeys(chain.Database(), header.ParentHash(nodeCtx))
+	if err != nil {
+		blake3pow.logger.Errorf("Failed to read colliding keys for block %s: %+v", header.ParentHash(nodeCtx).String(), err)
```

Comment:
can we continue after this ?

Reply:
This collidingKeys will very likely never exist. But the error is different from not existing I think, need to make sure of that